ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNode.java

```diff
@@ -244,12 +244,71 @@ private static void createSinkTopic(
         : Collections.emptyMap();
 
     kafkaTopicClient.createTopic(kafkaTopicName,
-                                 numberOfPartitions,
-                                 numberOfReplications,
-                                 config
+        numberOfPartitions,
+        numberOfReplications,
+        config
     );
   }
 
+  private static SourceTopicProperties getSourceTopicProperties(
+      final String kafkaTopicName,
+      final Map<String, Object> sinkProperties,
+      final KafkaTopicClient kafkaTopicClient,
+      final KsqlConfig ksqlConfig
+  ) {
+    final Map ksqlProperties = ksqlConfig.values();
+    if ((ksqlProperties.containsKey(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY)
+          && ksqlProperties.get(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY) != null)
+        || (ksqlProperties.containsKey(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY)
+          && ksqlProperties.get(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY) != null)) {
+      return getSinkTopicPropertiesLegacyWay(sinkProperties, ksqlConfig);
+    }
+    // Don't request topic properties from Kafka if both are set in WITH clause.
+    if (sinkProperties.containsKey(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY)
+        && sinkProperties.containsKey(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY)) {
+      return new SourceTopicProperties(
+          (Integer) sinkProperties.get(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY),
+          (Short) sinkProperties.get(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY)
+      );
+    }
+    final TopicDescription topicDescription = getSourceTopicPropertiesFromKafka(
+        kafkaTopicName,
+        kafkaTopicClient);
+
+    final int partitions = (Integer) sinkProperties.getOrDefault(
+        KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY,
+        topicDescription.partitions().size());
+    final short replicas = (Short) sinkProperties.getOrDefault(
+        KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY,
+        (short) topicDescription.partitions().get(0).replicas().size());
+
+    return new SourceTopicProperties(partitions, replicas);
+  }
+
+  private static SourceTopicProperties getSinkTopicPropertiesLegacyWay(
+      final Map<String, Object> outputProperties,
```

Comment:
nit: rename `outputProperties`  to `sinkProperties` will make this easier to understand.